src/envs/behavior_options.py

```diff
@@ -0,0 +1,1012 @@
+        return None
+
+    print("Position commanded is in collision or blocked!")
+    p.restoreState(state)
+    p.removeState(state)
+    print(f"PRIMITIVE: navigate to {obj.name} with params {pos_offset} fail")
+    return None
+
+
+#################
+
+# Grasp #
+
+
+# Sampler for grasp continuous params
+def grasp_obj_param_sampler(rng: Generator) -> np.ndarray:
+    """Sampler for grasp option."""
+    x_offset = (rng.random() * 0.4) - 0.2
+    y_offset = (rng.random() * 0.4) - 0.2
+    z_offset = rng.random() * 0.2
+    return np.array([x_offset, y_offset, z_offset])
+
+
+def get_delta_low_level_hand_action(  # type: ignore
+    env,
+    old_pos: Union[Sequence[float], np.ndarray],
+    old_orn: Union[Sequence[float], np.ndarray],
+    new_pos: Union[Sequence[float], np.ndarray],
+    new_orn: Union[Sequence[float], np.ndarray],
+) -> np.ndarray:
+    """Function to get low level actions from hand-movement plan."""
+    # First, convert the supplied orientations to quaternions
+    old_orn = p.getQuaternionFromEuler(old_orn)
+    new_orn = p.getQuaternionFromEuler(new_orn)
+
+    # Next, find the inverted position of the body (which we know shouldn't
+    # change, since our actions move either the body or the hand, but not
+    # both simultaneously)
+    body = env.robots[0].parts["right_hand"].parent.parts["body"]
+    inverted_body_new_pos, inverted_body_new_orn = p.invertTransform(
+        body.new_pos, body.new_orn)
+    # Use this to compute the new pose of the hand w.r.t the body frame
+    new_local_pos, new_local_orn = p.multiplyTransforms(
+        inverted_body_new_pos, inverted_body_new_orn, new_pos, new_orn)
+
+    # Next, compute the old pose of the hand w.r.t the body frame
+    inverted_body_old_pos = inverted_body_new_pos
+    inverted_body_old_orn = inverted_body_new_orn
+    old_local_pos, old_local_orn = p.multiplyTransforms(
+        inverted_body_old_pos, inverted_body_old_orn, old_pos, old_orn)
+
+    # The delta position is simply given by the difference between these
+    # positions
+    delta_pos = np.array(new_local_pos) - np.array(old_local_pos)
+
+    # Finally, compute the delta orientation
+    inverted_old_local_orn_pos, inverted_old_local_orn_orn = p.invertTransform(
+        [0, 0, 0], old_local_orn)
+    _, delta_orn = p.multiplyTransforms(
+        [0, 0, 0],
+        new_local_orn,
+        inverted_old_local_orn_pos,
+        inverted_old_local_orn_orn,
+    )
+
+    delta_trig_frac = 0
+    action = np.concatenate(  # type: ignore
+        [
+            np.zeros((10)),
+            np.array(delta_pos),
+            np.array(p.getEulerFromQuaternion(delta_orn)),
+            np.array([delta_trig_frac]),
+        ],
+        axis=0,
+    )
+
+    return action
+
+
+def grasp_obj_at_pos(  # type: ignore
+        env,
+        obj,
+        grasp_offset: np.ndarray,
+        rng: Generator = np.random.default_rng(23),
+) -> Union[None, Callable]:
```

Comment:
The `Union[None, X]` pattern in mypy is equivalent to the preferable `Optional[X]`

Reply:
Would it be easy to define the inputs and outputs of Callable here? You can look at how we do that everywhere else `git grep Callable`

Reply:
We run into the `BehaviorEnv` import issue again here: the proper type would be `Optional[Callable[[[State, BehaviorEnv], Tuple[Array, bool]]]]`, but `BehaviorEnv` won't get imported on CI because it's an igibson thing...